src/test_fixture/transport/routing.rs

```diff
@@ -165,28 +159,51 @@ impl Drop for Switch {
     }
 }
 
-#[derive(Default)]
-struct QueryCommandRouter {
-    routes: HashMap<QueryId, mpsc::Sender<CommandEnvelope>>,
+#[derive(Debug, thiserror::Error)]
+enum DispatchError {
+    #[error("No listeners subscribed for {command:?}")]
+    NoSubscribers { command: CommandEnvelope },
+    #[error("Failed to send {command:?}")]
+    SendFailed {
+        command: CommandEnvelope,
+        inner: BoxError,
+    },
 }
 
-impl QueryCommandRouter {
-    async fn route(&self, origin: HelperIdentity, query_id: QueryId, step: Step, payload: Vec<u8>) {
-        let sender = self
-            .routes
-            .get(&query_id)
-            .unwrap_or_else(|| panic!("No subscribers for {query_id:?}"));
-
-        sender
-            .send(CommandEnvelope {
-                origin: CommandOrigin::Helper(origin),
-                payload: TransportCommand::StepData(query_id, step, payload),
-            })
-            .await
-            .unwrap();
+impl From<mpsc::error::SendError<CommandEnvelope>> for DispatchError {
+    fn from(value: SendError<CommandEnvelope>) -> Self {
+        Self::SendFailed {
+            command: value.0,
+            inner: "channel closed".into(),
+        }
     }
+}
 
-    fn subscribe(&mut self, query_id: QueryId, sender: mpsc::Sender<CommandEnvelope>) {
-        assert!(self.routes.insert(query_id, sender).is_none());
+#[async_trait]
+trait Dispatcher {
```

Comment:
just curious -- what alternative implementers do you see for `Dispatcher`?

Reply:
likely only one implementation. I'll see if I can remove it

Reply:
I don't think I can because it is an extension trait, adding behavior to foreign type

Reply:
oh, yeah that makes sense